File: bot/greetings.py
import os
from discord.ext import commands
from asyncio import sleep
import discord
import logging


class Greetings(commands.Cog):
    def __init__(self, bot):
        # Variables to pre-load
        self.online_role = None
        self.current_guild = None

        self.logger = logging.getLogger('discord')
        self.logger.setLevel(logging.INFO)

        self.bot = bot

    async def cog_load(self):
        self.logger.info(f"Greetings.cog: LOADED!")

    # Member Greeting Embed
    @commands.Cog.listener()
    async def on_member_update(self, before, after):
        if not before.bot:
            if before.pending is True and after.pending is False:
                if self.online_role not in before.roles:  # Once onboarded and verified
                    await sleep(0.2)

                    the_guild: discord.Guild = await self.bot.fetch_guild(int(os.getenv("DEV_GUILD_ID")))
                    lounge_channel: discord.TextChannel = await the_guild.fetch_channel(
                        int(os.getenv("DEV_LOUNGE_ID")))

                    # Add "Online" role
                    await after.add_roles(the_guild.get_role(int(os.getenv("DEV_ONLINE_ROLE_ID"))))

                    welcome_embed = discord.Embed(
                        description=f"### <:Discord_Invite:1140057489941995650>"
                                    f"   Head over to <#{os.getenv('DEV_GET_ROLES_ID')}> to grab your roles"
                                    f"\n### <:Discord_Message_SpeakTTS:1140059207106826271>"
                                    f"   And remember to **`GIVE`** <#{os.getenv('DEV_FEEDBACK_ID')}> "
                                    f"__before__ you **`ASK`** for it!",
                        colour=13281772
                    )
                    welcome_embed.set_footer(text=f"If you have any questions, feel free to @ one of our moderators!")

                    await lounge_channel.send(f"**Hello {before.mention} and welcome to ChillSynth!**")
                    await lounge_channel.send(embed=welcome_embed)


async def setup(bot):
    await bot.add_cog(Greetings(bot))

File: internal/greetings/greetings.go
package greetings

import (
	"chillbot/internal/bot"
	"chillbot/internal/module"
	"fmt"
	"slices"

	"github.com/bwmarrin/discordgo"
)

type GreetingsModule struct {
	module.CommonDeps
}

func (m *GreetingsModule) Load(deps *module.CommonDeps) {
	m.Bot = deps.Bot
	m.Logger = deps.Logger
	m.Config = deps.Config

	bot.AddHandler(m.Bot, m.GreetVerifiedUser)
}

func (m *GreetingsModule) GreetVerifiedUser(g *discordgo.GuildMemberUpdate) error {
	if g.Member.User.Bot || g.BeforeUpdate == nil || g.GuildID != m.Config.GuildID {
		return nil
	}
	if m.hasUserJustBeenVerified(g) {
		err := m.Bot.Discord.GuildMemberRoleAdd(g.Member.GuildID, g.Member.User.ID, m.Config.Roles["online_role"])
		if err != nil {
			return err
		}

		_, err = m.Bot.Discord.ChannelMessageSend(m.Config.LoungeChannelID,
			fmt.Sprintf("**Hello %s and welcome to ChillSynth!**", g.Mention()))
		if err != nil {
			return err
		}

		_, err = m.Bot.Discord.ChannelMessageSendEmbed(m.Config.LoungeChannelID, &discordgo.MessageEmbed{
			Description: fmt.Sprintf(
				"### <:Discord_Invite:1140057489941995650> Head over to <#%s> to grab your roles \n"+
					"### <:Discord_Message_SpeakTTS:1140059207106826271> "+
					"And remember to **`GIVE`** <#%s> __before__ you **`ASK`** for it!",
				m.Config.LoungeChannelID,
				m.Config.FeedbackChannelID),
			Color: 13281772,
			Footer: &discordgo.MessageEmbedFooter{
				Text: "If you have any questions, feel free to @ one of our moderators!",
			},
		})
		if err != nil {
			return err
		}
	}
	return nil
}

func (m *GreetingsModule) hasUserJustBeenVerified(g *discordgo.GuildMemberUpdate) bool {
	return g.BeforeUpdate.Pending && !g.Pending && !slices.Contains(g.BeforeUpdate.Roles, m.Config.Roles["online_role"])
}
